design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,192 @@
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce two new attributes: `parametrize` and `case`.
+
+- `parametrize` - defines a parametrized test.
+- `case` - defines a specific test case
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Simple Arguments
+
+```cairo
+#[parametrize]
+#[case(1, 3)]
+#[case(3, 5)]
+#[test]
+fn my_test(a: felt252, b: u32) {
+    // case a = 1, b = 3
+    // case a = 3, b = 5
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 1;
+}
+
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 3;
+  let b: u32 = 5;
+}
+```
+
+The explicit type annotations for each argument are necessary.
+This way, if the user provides invalid value for the type defined, the compiler will throw a relevant error.
+
+### Parametrization With Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture() -> MyStruct {
+    MyStruct { a: 1, b: array![2, 3] }
+}
+
+#[parametrize]
+#[case(1, my_fixture())]
+#[case(3, my_fixture())]
+#[test]
+fn my_test(a: felt252, b: MyStruct) {
+    // case a = 1, b = MyStruct { a: 1, b: [2, 3] }
+    // case a = 3, b = MyStruct { a: 1, b: [2, 3] }
+    // ...
+}
+```
+
+For fixtures, we would generate code in the exactly same manner as for simple arguments.
+
+```cairo
+// ...
+#[test]
+fn my_test_1() {
+    let a = 1;
+    let b = my_fixture();
+    // ...
+}
+```
+
+### Parametrized Fixtures
+
+```cairo
+struct MyStruct {
+    a: felt252, 
+    b: Array<felt252>
+}
+
+fn my_fixture(a: felt252, b: felt252, c: felt252) -> MyStruct {
+    MyStruct { a: a, b: array![b, c] }
+}
+
+#[parametrize]
+#[case(1, my_fixture(2, 3, 4))]
+#[case(3, my_fixture(5, 6, 7))]
+#[test]
+fn my_test(a: felt252, b: MyStruct) {
+    // case a = 1, b = MyStruct { a: 2, b: [3, 4] }
+    // case a = 3, b = MyStruct { a: 5, b: [6, 7] }
+    // ...
+}
+```
+
+This could be handled in a very similar manner as the [non-parametrized fixture case](#parametrization-with-fixtures):
+
+```cairo
+// ...
+#[test]
+fn my_test_generated(a: felt252) {
+    let a = 1;
+    let b = my_fixture(2, 3, 4);
+    // ...
+}
+```
+
+### Parametrizing With Complex Types
+
+While it is technically possible to use complex types in attributes instead of defining fixtures, handling of them in
+code would be quite problematic based on my limited research.
```

Comment:
I don't understand why this would be more problematic than fixture composing, constructors are not much different from functions. 
Imo this syntax makes more sense and we should focus on.

Reply:
We could look into using cairo parser to parse this syntax. Imo that should be a goal of this design doc.

Reply:
From my limited research, annotations aren't really type-checked, at least not by default. You get an reference to a node in AST with whatever you have in there and have to handle it. I don't think it is easy or obvious to handle complex structures in this so that's why I'm not proposing a solution like this.

Only allowing simple values or calls to functions (no support for nested calls etc.) seems easy enough to implement however.

Reply:
> Only allowing simple values or calls to functions (no support for nested calls etc.) seems easy enough to implement however.

I don't think this is based on anything other than your intuition. Please explain in the document what you have tried in regards of structs and where you ran into problems. Explain how this checking will work, what exact restrictions will be in place.

Reply:
I've added an explanation to the document, though I'm not sure if we should be including things like that.

Reply:
Not relevant anymore